File: extensions/panache/mongodb-panache-kotlin/runtime/src/main/java/io/quarkus/mongodb/panache/kotlin/reactive/runtime/ReactivePanacheQueryImpl.java
package io.quarkus.mongodb.panache.kotlin.reactive.runtime;

import java.util.List;

import org.bson.conversions.Bson;

import com.mongodb.ReadPreference;
import com.mongodb.client.model.Collation;

import io.quarkus.mongodb.panache.kotlin.reactive.ReactivePanacheQuery;
import io.quarkus.mongodb.panache.reactive.runtime.CommonReactivePanacheQueryImpl;
import io.quarkus.mongodb.reactive.ReactiveMongoCollection;
import io.quarkus.panache.common.Page;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;

@SuppressWarnings("unchecked")
public class ReactivePanacheQueryImpl<Entity> implements ReactivePanacheQuery<Entity> {
    private final CommonReactivePanacheQueryImpl<Entity> delegate;

    ReactivePanacheQueryImpl(ReactiveMongoCollection<? extends Entity> collection, Bson mongoQuery, Bson sort) {
        this.delegate = new CommonReactivePanacheQueryImpl<>(collection, mongoQuery, sort);
    }

    private ReactivePanacheQueryImpl(CommonReactivePanacheQueryImpl<Entity> delegate) {
        this.delegate = delegate;
    }

    @Override
    public <T> ReactivePanacheQuery<T> project(Class<T> type) {
        return new ReactivePanacheQueryImpl<>(delegate.project(type));
    }

    @Override
    @SuppressWarnings("unchecked")
    public ReactivePanacheQuery<Entity> page(Page page) {
        delegate.page(page);
        return this;
    }

    @Override
    public ReactivePanacheQuery<Entity> page(int pageIndex, int pageSize) {
        return page(Page.of(pageIndex, pageSize));
    }

    @Override
    public ReactivePanacheQuery<Entity> nextPage() {
        delegate.nextPage();
        return this;
    }

    @Override
    public ReactivePanacheQuery<Entity> previousPage() {
        delegate.previousPage();
        return this;
    }

    @Override
    public ReactivePanacheQuery<Entity> firstPage() {
        delegate.firstPage();
        return this;
    }

    @Override
    public Uni<ReactivePanacheQuery<Entity>> lastPage() {
        Uni<CommonReactivePanacheQueryImpl<Entity>> uni = delegate.lastPage();
        return uni.map(q -> this);
    }

    @Override
    public Uni<Boolean> hasNextPage() {
        return delegate.hasNextPage();
    }

    @Override
    public boolean hasPreviousPage() {
        return delegate.hasPreviousPage();
    }

    @Override
    public Uni<Integer> pageCount() {
        return delegate.pageCount();
    }

    @Override
    public Page page() {
        return delegate.page();
    }

    @Override
    public ReactivePanacheQuery<Entity> range(int startIndex, int lastIndex) {
        delegate.range(startIndex, lastIndex);
        return this;
    }

    @Override
    public ReactivePanacheQuery<Entity> withCollation(Collation collation) {
        delegate.withCollation(collation);
        return this;
    }

    @Override
    public ReactivePanacheQuery<Entity> withReadPreference(ReadPreference readPreference) {
        delegate.withReadPreference(readPreference);
        return this;
    }

    @Override
    public Uni<Long> count() {
        return delegate.count();
    }

    @Override
    public Uni<List<Entity>> list() {
        return delegate.list();
    }

    @Override
    public Multi<Entity> stream() {
        return delegate.stream();
    }

    @Override
    public Uni<Entity> firstResult() {
        return delegate.firstResult();
    }

    @Override
    @SuppressWarnings("unchecked")
    public Uni<Entity> singleResult() {
        return delegate.singleResult();
    }
}

File: extensions/panache/mongodb-panache-kotlin/runtime/src/main/kotlin/io/quarkus/mongodb/panache/kotlin/reactive/runtime/ReactivePanacheQueryImpl.kt
package io.quarkus.mongodb.panache.kotlin.reactive.runtime

import io.quarkus.mongodb.panache.common.reactive.runtime.CommonReactivePanacheQueryImpl
import io.quarkus.mongodb.reactive.ReactiveMongoCollection
import org.bson.conversions.Bson
import io.smallrye.mutiny.Uni
import com.mongodb.ReadPreference
import com.mongodb.client.model.Collation
import io.quarkus.mongodb.panache.kotlin.reactive.ReactivePanacheQuery
import io.quarkus.panache.common.Page
import io.smallrye.mutiny.Multi

class ReactivePanacheQueryImpl<Entity> : ReactivePanacheQuery<Entity> {
    private val delegate: CommonReactivePanacheQueryImpl<Entity>

    internal constructor(collection: ReactiveMongoCollection<out Entity>?, mongoQuery: Bson?, sort: Bson?) {
        delegate = CommonReactivePanacheQueryImpl(collection, mongoQuery, sort)
    }

    private constructor(delegate: CommonReactivePanacheQueryImpl<Entity>) {
        this.delegate = delegate
    }

    override fun <T> project(type: Class<T>): ReactivePanacheQuery<T> {
        return ReactivePanacheQueryImpl(delegate.project(type))
    }

    override fun page(page: Page): ReactivePanacheQuery<Entity> {
        delegate.page<Entity>(page)
        return this
    }

    override fun page(pageIndex: Int, pageSize: Int): ReactivePanacheQuery<Entity> {
        return page(Page.of(pageIndex, pageSize))
    }

    override fun nextPage(): ReactivePanacheQuery<Entity> {
        delegate.nextPage<Entity>()
        return this
    }

    override fun previousPage(): ReactivePanacheQuery<Entity> {
        delegate.previousPage<Entity>()
        return this
    }

    override fun firstPage(): ReactivePanacheQuery<Entity> {
        delegate.firstPage<Entity>()
        return this
    }

    override fun lastPage(): Uni<ReactivePanacheQuery<Entity>> {
        val uni: Uni<CommonReactivePanacheQueryImpl<Entity>> = delegate.lastPage()
        return uni.map { this }
    }

    override fun hasNextPage(): Uni<Boolean> {
        return delegate.hasNextPage()
    }

    override fun hasPreviousPage(): Boolean {
        return delegate.hasPreviousPage()
    }

    override fun pageCount(): Uni<Int> {
        return delegate.pageCount()
    }

    override fun page(): Page {
        return delegate.page()
    }

    override fun range(startIndex: Int, lastIndex: Int): ReactivePanacheQuery<Entity> {
        delegate.range<Entity>(startIndex, lastIndex)
        return this
    }

    override fun withCollation(collation: Collation): ReactivePanacheQuery<Entity> {
        delegate.withCollation<Entity>(collation)
        return this
    }

    override fun withReadPreference(readPreference: ReadPreference): ReactivePanacheQuery<Entity> {
        delegate.withReadPreference<Entity>(readPreference)
        return this
    }

    override fun count(): Uni<Long> {
        return delegate.count()
    }

    override fun list(): Uni<List<Entity>> {
        return delegate.list()
    }

    override fun stream(): Multi<Entity> {
        return delegate.stream()
    }

    override fun firstResult(): Uni<Entity?> {
        return delegate.firstResult()
    }

    override fun singleResult(): Uni<Entity?> {
        return delegate.singleResult()
    }
}
